### DISCUSSION TITLE
Data List: Conditional Data types

### DISCUSSION THREAD

**Discussion by AaronSadlerUK** [leekelleher/umbraco-contentment#131] [Idea]
It would be cool if we could use https://github.com/KOBENDigital/ConditionalDisplayers with the contentment data types 😁

**Comment by leekelleher** [comment]
I had a look at how Conditional Displayers works, to see if there were any generic hooks to work with, but it doesn't look possible - well, not without re-designing Data List to be dependent on Conditional Displayers, _(which I'm not interested in doing)._

Looking at how Conditional Displayers works, in the dropdown one, when the value changes, it triggers this event... https://github.com/KOBENDigital/ConditionalDisplayers/blob/2.3.0/src/App_Plugins/ConditionalDisplayers/propertyeditors/dropdownFlexible/dropdownFlexible.controller.js#L25 ...making a call to `cdSharedLogic.displayProps()` https://github.com/KOBENDigital/ConditionalDisplayers/blob/2.3.0/src/App_Plugins/ConditionalDisplayers/propertyeditors/cdSharedLogic.js#L9

It appears that each of the dropdown values has options to set which property aliases can be shown and hidden. I'm unsure how that would work with Data List, nor how it would work across all the Data List editors (e.g. Item Picker).

One approach could be to make use of Umbraco's `eventsService.emit` feature? Umbraco core uses it on the `umb-toggle` directive (used in the Checkbox property-editor). e.g. https://github.com/umbraco/Umbraco-CMS/blob/release-8.16.0/src/Umbraco.Web.UI.Client/src/common/directives/components/buttons/umbtoggle.directive.js#L85

Then, say for the Data List's dropdown editor, there is already an event for when the value changes, e.g. https://github.com/leekelleher/umbraco-contentment/blob/2.1.0/src/Umbraco.Community.Contentment/DataEditors/DropdownList/dropdown-list.js#L48

I've used the `eventsService.emit` for the Data List preview (on the Data Type).  It's set up on another one of the Contentment editors (an internal one called Configuration Editor), that broadcasts when a value has changed... e.g. https://github.com/leekelleher/umbraco-contentment/blob/2.1.0/src/Umbraco.Community.Contentment/DataEditors/ConfigurationEditor/configuration-editor.js#L188
Then in the Data List preview editor, it is notified of any changes and updates accordingly. e.g. https://github.com/leekelleher/umbraco-contentment/blob/2.1.0/src/Umbraco.Community.Contentment/DataEditors/DataList/data-list.preview.js#L49

Theoretically, the conditional displayers could be configured elsewhere (rather than the property-editor/data-type itself) and the `eventsService` could be used to trigger the conditions.  As for where that is configured, I've no idea, this is all purely brainstorming.

**Reply by skartknet** [reply]
Hey guys, are we talking here about displaying properties that use Contentment editors conditionally? Or using the Data List values to populate the conditional displayers editors?

**Reply by AaronSadlerUK** [reply]
So i would like to use the contentment data list as it supports key/values from various data sources.

But i would like to hide other properties depending on the selection

**Reply by solarpolar** [reply]
Were you guys ever able to work out any conditional options?  I'm trying to figure how to use the value picked from my first datalist field, to then conditionally display further differing datalist fields, all depending on the first field's value.  

Just trying to figure out if there's a way, or if I have to use bespoke Property Editors.

**Reply by leekelleher** [reply]
@solarpolar I didn't do anything more on this (specifically for the Data List editors). I did do some experimental prototyping of a Cascading editor, but didn't make the source-code public... however it was very similar to the (internal) Cascading DropdownList editor that is used on the Data List's Enum data-source configuration.

See here for details: https://github.com/leekelleher/umbraco-contentment/discussions/202#discussioncomment-2294685 (however limited to dropdown editors)

**Reply by JohanReitsma83** [reply]
@solarpolar did you work on it or have a solution? We are also looking into a custom data list picker where we want to hide certain properties if a value is selected. Maybe we can do it together or help @leekelleher ?

**Comment by skartknet** [comment]
I have started working on creating a couple of Data List Editors that use Conditional Displayers. I will update you once they are done so you can give some feedback.

**Reply by JohanReitsma83** [reply]
@skartknet any update?

**Reply by skartknet** [reply]
I haven't had time to keep looking. Follow this thread: https://github.com/leekelleher/umbraco-contentment/discussions/241#discussioncomment-3302687

**Reply by leekelleher** [reply]
I prototyped an alternative approach during the last [Umbracalong](https://www.meetup.com/umbracalong/), where the conditional logic was in a separate property-editor, setting watchers on the source editor's value and switching the show/hide on the target editor(s). I put the code in a gist: https://gist.github.com/leekelleher/afb683adad38527388354b612621ab43.

Keep in mind that this was a quick-n-dirty proof-of-concept, very much hacking around. It seemed to work okay for single value editors, but unpredictable for multi-value editors.

I (currently) have no interest in pursuing this myself - as I scratched my developer-itch, to see if it would work (or not). Feel free to take/fork/modify the code from the gist.

**Reply by skartknet** [reply]
Thanks for that Lee, I'll take a look, hopefully soon!

**Reply by skartknet** [reply]
The problem I'm having (and that's described on the other discussion) is that I can't get the contentment's data source values on the Configuration Field. With using this gist we still would have the same problem as we would need to configure it entering text values.

**Reply by leekelleher** [reply]
@skartknet The code in the gist would be for a separate property-editor, it'd have no context of the Data List (or any other editor), only what you'd want the conditional values to be. I hadn't considered how the data-type configuration would work out, I just wanted to prototype the content-editor parts.

(In the other thread, I'm still unsure why that value was `null` for you. I couldn't reproduce that issue. 😕 ref: https://github.com/leekelleher/umbraco-contentment/discussions/241#discussioncomment-3309212)

**Comment by leekelleher** [comment]
Closing this off, (since it's been a few years).

This may be viable with the new backoffice (via the [Property Dataset Context](https://docs.umbraco.com/umbraco-cms/extend-your-project/backoffice-extensions/contexts/property-dataset-context)), but it's not something that I'd look to bake into Contentment's Data List.

I'm happy to re-open this if the Conditional Displayers package (or any other package) needs collaborate with me to make Data List work with it.
